src/ast/passes/printer.cpp

```diff
@@ -81,513 +278,1067 @@ void Printer::visit(String &string)
 {
-  std::string indent(depth_, ' ');
-  out_ << indent << "builtin: " << builtin.ident << type(builtin.builtin_type)
-       << std::endl;
+  // Does not have a syntactic representation. We should never get
+  // here during a normal traversal, but we can just emit "none".
+  return Buffer().text("none");
 }
 
-void Printer::visit(Identifier &identifier)
+Buffer Formatter::visit(Builtin& builtin)
 {
-  std::string indent(depth_, ' ');
-  out_ << indent << "identifier: " << identifier.ident
-       << type(identifier.ident_type) << std::endl;
+  return Buffer().text(builtin.ident);
 }
 
-void Printer::visit(Call &call)
+Buffer Formatter::visit(Identifier& identifier)
 {
-  std::string indent(depth_, ' ');
-  out_ << indent << "call: " << call.func << type(call.return_type)
-       << std::endl;
-
-  ++depth_;
-  visit(call.vargs);
-  --depth_;
+  return Buffer().text(identifier.ident);
 }
 
-void Printer::visit(Sizeof &szof)
+template <typename T>
+bool is_comparison(const T& expr)
 {
-  std::string indent(depth_, ' ');
-  out_ << indent << "sizeof: " << std::endl;
-
-  ++depth_;
-  visit(szof.record);
-  --depth_;
+  if constexpr (std::is_same_v<Expression, T>) {
+    auto* binop = expr.template as<Binop>();
+    if (!binop) {
+      return false;
+    }
+    switch (binop->op) {
+      case Operator::EQ:
+      case Operator::NE:
+      case Operator::LE:
+      case Operator::GE:
+      case Operator::LT:
+      case Operator::GT:
+        return true;
+      default:
+        return false;
+    }
+  }
+  return false;
 }
 
-void Printer::visit(Offsetof &offof)
+template <typename U>
+Buffer Formatter::format_list(std::vector<U>& exprs,
+                              const std::string& sep,
+                              MetadataIndex& metadata,
+                              size_t max_width,
+                              bool bare)
 {
-  std::string indent(depth_, ' ');
-  out_ << indent << "offsetof: " << std::endl;
+  // We need to carefully extract the metadata for all the arguments, and
+  // position them correctly. Depending on the width, we will do either
+  // multiline in inline.
+  std::vector<Buffer> args;
+  size_t total_width = 0;
+  for (size_t i = 0; i < exprs.size(); i++) {
+    auto arg = format(exprs[i],
+                      metadata,
+                      (i < exprs.size() - 1) ? max_width - 1 : max_width,
+                      bare || is_comparison(exprs[i]));
+    if (i < exprs.size() - 1) {
+      arg = arg.text(sep);
+    }
+    total_width += arg.width();
+    args.emplace_back(std::move(arg));
+  }
+  if (total_width < max_width) {
+    // Join into a single line.
+    auto single_line_args = Buffer();
+    for (size_t i = 0; i < exprs.size(); i++) {
+      single_line_args = single_line_args.append(std::move(args[i]));
+    }
+    return single_line_args;
+  } else {
+    // Find the most aesthetically pleasing arrangement by minimizing
+    // the relative gap between the widest and narrowest lines.
```

Comment:
Oh, wow, this is some effort to make it look nice!

Reply:
Doing this left me with an appreciation for how hard the problem generally is :) Without this a lot of stuff just looks weird.